### PR TITLE
Various fixes based on random data experiments

### PR DESCRIPTION
**Pull request by sabraaap**
-account for cases where the agent is successful at the last time step
-some adjustments to the existing logic that allows for training with full episodes
-including last episode step for RTG calculation
-evaluating sample efficiency on training seeds during training, IID generalisation (val seeds) and OOD generalisation (test seeds) at the end of training
-using dataset `len` attribute to set number of training samples (1 sample = 1 sub-episode / 1 action prediction), and setting this to be either the number of episodes in the dataset or specifying a (smaller) number of samples

### CODE REVIEW

File: src/collator/collator.py [hunk 1]
```diff
@@ -112,7 +115,11 @@ def embed_sentences(self, sentences, type):
 
     # helper func to pad 2D or 3D numpy array along axis 1
     def _pad(self, x, pad_width=None, before=True, val=0):
-        pad_width = pad_width or max(self.context_length - x.shape[1], 0)
+        pad_width = pad_width or (
```
Comment (sabraaap) [inline src/collator/collator.py:118]: adjusted padding for full length episodes

File: src/collator/collator.py [hunk 2]
```diff
@@ -202,7 +209,8 @@ def _sample_batch(self, batch_size, random_start=True, full=False, train=True):
 
         # if we're in training mode, update the sample counter
         if train:
-            self.samples_processed += self._count_samples_processed(batch)
+            # self.samples_processed += self._count_samples_processed(batch)
+            self.samples_processed += batch_size
```
Comment (sabraaap) [inline src/collator/collator.py:213]: do not count context steps as samples

File: src/dataset/custom_dataset.py [hunk 3]
```diff
@@ -35,9 +35,10 @@ def __init__(self, args):
         self.total_episodes = 0
         self.seed_finder = SeedFinder()
         self.level = self.args["level"]
+        self.category = self._get_category()
         self.train_config, self.test_configs = self._get_configs()
         self.train_seeds = self._get_train_seeds()
-        self.category = self._get_category()
+        self.max_steps = self._get_level_max_steps()
```
Comment (sabraaap) [inline src/dataset/custom_dataset.py:41]: setting max steps on init as this is being used repeatedly throughout the code and needs to be accessible upon loading

File: src/dataset/custom_dataset.py [hunk 4]
```diff
@@ -421,7 +424,7 @@ def sample_episode_indices(self, num_eps, dist="uniform"):
             probs = 1 / self.episode_lengths
         else:
             probs = np.ones(self.num_episodes)
-        probs /= np.sum(probs)
+        probs = probs / np.sum(probs)
```
Comment (sabraaap) [inline src/dataset/custom_dataset.py:427]: this seems to break when using the /= notation with "length" (due to one being float, the other int)

File: src/dataset/custom_dataset.py [hunk 5]
```diff
@@ -438,22 +441,24 @@ def sample_episode(
...
+
         s = self.shard.observations[start:end]
         s = normalise(s).reshape(1, -1, self.state_dim)
 
         a = self.shard.actions[start:end]
         a = to_one_hot(a, self.act_dim).reshape(1, -1, self.act_dim)
 
         rtg = discounted_cumsum(
-            self.shard.rewards[start : self.episode_ends[ep_idx]], gamma=gamma
+            self.shard.rewards[start : self.episode_ends[ep_idx] + 1], gamma=gamma
```
Comment (sabraaap) [inline src/dataset/custom_dataset.py:461]: so we're including the actual episode end step in the calculation

File: src/dataset/custom_dataset.py [hunk 6]
```diff
@@ -579,7 +584,7 @@ def callback(exps, logs, seeds):
         return self
 
     def __len__(self):
-        return self.num_shards * self.eps_per_shard
+        return min(self.num_shards * self.eps_per_shard, self.args["num_samples"])
```
Comment (sabraaap) [inline src/dataset/custom_dataset.py:587]: this is just a suggestion - decided for this as it worked with the full length episodes

File: src/env/feedback_env.py [hunk 7]
```diff
@@ -79,8 +79,14 @@ def step(self, action):
 
         # check if max steps reached
         self.steps_taken += 1
-        if self.max_steps is not None and self.steps_taken >= self.max_steps:
+        if (
+            self.max_steps is not None
+            and self.steps_taken >= self.max_steps
+            and not terminated
```
Comment (sabraaap) [inline src/env/feedback_env.py:85]: for cases where episodes are terminated at the last step

File: src/ppo/ppo_agent.py [hunk 8]
```diff
@@ -57,7 +57,7 @@ def __init__(
             if env_name in ["BabyAI-OpenTwoDoors-v0", "BabyAI-OpenRedBlueDoors-v0"]
             else 1,
             "text": True,
-            "argmax": True,
+            "argmax": False,
```
Comment (sabraaap) [inline src/ppo/ppo_agent.py:60]: just for consistency

File: src/trainer/evaluator.py [hunk 9]
```diff
@@ -174,36 +173,61 @@ def on_train_end(
         **kwargs,
     ):
         self._plot_loss(state)
-        self._run_eval_and_plot(model, state, eval_type="generalisation")
+        self._run_eval_and_plot(model, state, eval_type="iid_generalisation")
```
Comment (sabraaap) [inline src/trainer/evaluator.py:176]: now using val seeds for end-of-training generalisation evaluation

File: src/env/feedback_env.py (hunk 7 above)
Comment (maxtaylordavies) [reply src/env/feedback_env.py:85]: v good catch thanks

**Review by maxtaylordavies — APPROVED** [review]
looks good, nice job! 🚀💪

**Pull request merged**